src/fragment-generation-utils.js

```diff
@@ -1141,6 +1152,47 @@ const getLastNodeForBlockSearch = (range) => {
   return node;
 };
 
+/**
+ * Finds the first visible text node within a given range.
+ * @param {Range} range - range in which to find the first visible text node
+ * @returns {Node} - first visible text node within |range| or null if there are
+ * no visible text nodes within |range|
+ */
+const getFirstTextNode = (range) => {
+  // Check if first node in the range is a visible text node.
+  const firstNode = getFirstNodeForBlockSearch(range);
+  if (isText(firstNode) && fragments.internal.isNodeVisible(firstNode)) {
+    return firstNode;
+  }
+
+  // First node is not visible text, use a tree walker to find the first visible
+  // text node.
+  const walker = fragments.internal.makeTextNodeWalker(range);
+  walker.currentNode = firstNode;
+
+  return walker.nextNode();
+};
+
+/**
+ * Finds the last visible text node within a given range.
+ * @param {Range} range - range in which to find the last visible text node
+ * @returns {Node} - last visible text node within |range| or null if there are
+ * no visible text nodes within |range|
+ */
+const getLastTextNode = (range) => {
+  // Check if last node in the range is a visible text node.
+  const lastNode = getLastNodeForBlockSearch(range);
+  if (isText(lastNode) && fragments.internal.isNodeVisible(lastNode)) {
+    return lastNode;
+  }
+
+  // Last node is not visible text, traverse the range backwards to find the
+  // last visible text node.
+  const walker = fragments.internal.makeTextNodeWalker(range);
+
+  return fragments.internal.backwardTraverse(walker, new Set());
```

Comment:
It seems odd to me that we use `nextNode` on line 1173. but `backwardTraverse` here. Should we switch to `forwardTraverse` above or `walker.previousNode` here, or is there a reason these need to be different?

Reply:
In the start side nestNode works because what we want is the next node that passes the filter ( text node inside the range) in DOM order (pre order tree traversal). 

Using the same approach on the end side ( a reversed pre order traversal) would require to start from the "rightmost" node inside the range, which translates to the rightmost node in the range's endContainer subtree. After getting that node, we could do walker.previousNode ( if the initial node is not visible text ) to find the first text node in the reversed pre order traversal. 
 
Backwards traverse, does the trick without having to find the initial rightmost node because given any starting node it will first go down the rightmost branch of the node's subtree emitting nodes that pass the filter and then traverse the tree backwards.

Reply:
sgtm